lib/acme/resources/challenges/tlssni01.rb

```diff
@@ -0,0 +1,24 @@
+class Acme::Resources::Challenges::TLSSNI01 < Acme::Resources::Challenges::Base
+
+  def hostname
+    hd = crypto.digest.hexdigest(authorization_key)
+    "#{hd[0..31]}.#{hd[32..64]}.acme.invalid"
+  end
+
+  def temporary_cert
+    @temporary_cert ||= client.new_self_signed_certificate self.hostname 
```

Comment:
As said I think the generation code should go into the challenge code for now, see above for some reasoning about that.